src/mctpd.c

```diff
@@ -2836,21 +2836,42 @@ static int method_learn_endpoint(sd_bus_message *call, void *data,
 static int query_peer_properties(struct peer *peer)
 {
 	int rc;
+	const int max_retries = 4;
 
 	rc = query_get_peer_msgtypes(peer);
 	if (rc < 0) {
 		// Warn here, it's a mandatory command code.
 		// It might be too noisy if some devices don't implement it.
 		warnx("Error getting endpoint types for %s. Ignoring error %d %s",
-		      peer_tostr(peer), rc, strerror(-rc));
+			peer_tostr(peer), rc, strerror(-rc));
+		if (rc == -ETIMEDOUT) {
+			for (int i = 0; i < max_retries; i++) {
+				if (peer->ctx->verbose)
+					warnx("Retrying to get endpoint types for %s. Attempt %d",
+						peer_tostr(peer), i + 1);
+				rc = query_get_peer_msgtypes(peer);
```

Comment:
With this change, we now have duplicate calls to `query_get_peer_msgtypes()`. If you include the first call in the retry loop you can eliminate this.

No need for including the first `warnx` warning in the timeout case.

Same with query_get_peer_uuid() below.